### PR TITLE
fix to: cascader's multiple determines the declaration of onChange

### PR DESCRIPTION
**Pull request by babycannotsay**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
#33153 

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  cascader's multiple determines the declaration of onChange  |
| 🇨🇳 Chinese |  cascader组件的multiple属性决定onChange的声明类型        |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/cascader/index.tsx [hunk 1]
```diff
@@ -74,15 +74,25 @@ const defaultSearchRender: ShowSearchType['render'] = (inputValue, path, prefixC
...
+type SingleCascaderProps = Omit<
+  Extract<RcCascaderProps, { checkable?: false }>,
+  'checkable' | 'options'
+> & { multiple?: false };
+type MultipleCascaderProps = Omit<
+  Extract<RcCascaderProps, { checkable: true | React.ReactNode }>,
+  'checkable' | 'options'
+> & { multiple: true };
+
+type UnionCascaderProps = SingleCascaderProps | MultipleCascaderProps;
```
Comment (babycannotsay) [inline components/cascader/index.tsx:86]: 是否需要从rc-cascader中导出这2个属性声明。

File: components/cascader/index.tsx [hunk 2]
```diff
@@ -74,15 +74,25 @@ const defaultSearchRender: ShowSearchType['render'] = (inputValue, path, prefixC
   return optionList;
 };
 
-export interface CascaderProps<DataNodeType>
-  extends Omit<RcCascaderProps, 'checkable' | 'options'> {
+type SingleCascaderProps = Omit<
```
Comment (babycannotsay) [inline components/cascader/index.tsx:77]: 必须在用联合类型前用Omit。
`type P = { x: string, z: string } | { x: number }`
`type C = Pick<P, 'x'> // { x: string | number }`

File: components/cascader/index.tsx [hunk 3]
```diff
@@ -74,15 +74,25 @@ const defaultSearchRender: ShowSearchType['render'] = (inputValue, path, prefixC
...
+  Extract<RcCascaderProps, { checkable?: false }>,
+  'checkable' | 'options'
+> & { multiple?: false };
+type MultipleCascaderProps = Omit<
+  Extract<RcCascaderProps, { checkable: true | React.ReactNode }>,
```
Comment (babycannotsay) [inline components/cascader/index.tsx:82]: rc-cascader包中，`{ checkable: true | React.ReactNode }`加了React.ReactNode会使得引用组件时，如果checkable为false，onChange的声明为onSingleChange | onMultipleChange | undefined
可以看[简单demo](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wFgAoCq8mATzCTgGEUBnNFAEySgAUcwrOAF44AbwpwpcNAAskaANYoARgBskAfgBccTCjWskAbknSQAVzUxgYDTr0Gjp8tLgQAdk1koPAcyRdVhgoYH8XAF84AB8pCVdpOQVldUC4EItGWOR0GAA6HIwAOQhuFzdLa1t7XQyTMylPb18A3Q8LEBUeSJc0T2CZXRZ2Lh5+CEERcThKmzs0-UMkABp3Lx9-RgiKPo8BziG2Dm4+ASFRMRl5JVUNXUWjVaaNgLhtynIgA)

File: components/cascader/index.tsx (hunk 1 above)
Comment (afc163) [reply components/cascader/index.tsx:86]: 最好从 rc-cascader 拿。

Comment (babycannotsay) [reply components/cascader/index.tsx:86]: 目前是没有导出的，需要去再提一份pr吗

Comment (babycannotsay) [reply components/cascader/index.tsx:87]: 已从rc-cascader中获取

**Comment by afc163** [comment]
https://github.com/react-component/cascader/releases/tag/v2.3.2

**Comment by afc163** [comment]
@babycannotsay ping

**Comment by babycannotsay** [comment]
files change打不开了，我这边重新fork提一个pr，这个先关了

**Comment by MadCcc** [comment]
github 的问题，好多都打不开了，建议等等

**Comment by babycannotsay** [comment]
emm，我原来以为是merge之后代码量太大导致的，原fork代码已经删了重新fork了。

**Pull request closed without merging**